cpp/src/arrow/util/async_generator.h

```diff
@@ -336,6 +345,68 @@ class ReadaheadGenerator {
   std::queue<Future<T>> readahead_queue_;
 };
 
+template <typename T>
+class PushGenerator {
```

Comment:
One issue with this model is that there is no back-pressure.  In the other generators they have bounded queuing so there is no way a slow consumer / fast producer can end up filling up RAM (the producer will be stopped while the consumer catches up).  Depending on use this may not be a problem (e.g. I think you're using this for very small file info objects which all fit in memory easily).

Reply:
Right. The use case I'm thinking about is async `GetFileInfo` for networked filesystems (such as S3). The aim is to cover latencies, but no significant CPU or RAM consumption should be involved. Also, in this use case, the producer would potentially run on multiple threads.

I'll add a comment.